core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1890,47 +1890,58 @@ class SparkContext(config: SparkConf) extends Logging {
             throw new IllegalArgumentException(
               s"Directory ${path} is not allowed for addJar")
           }
-          path
+          Seq(path)
         } catch {
           case NonFatal(e) =>
             logError(s"Failed to add $path to Spark environment", e)
-            null
+            Nil
         }
       } else {
-        path
+        Seq(path)
       }
     }
 
     if (path == null || path.isEmpty) {
       logWarning("null or empty path specified as parameter to addJar")
     } else {
-      val key = if (path.contains("\\") && Utils.isWindows) {
+      val (keys, schema) = if (path.contains("\\") && Utils.isWindows) {
         // For local paths with backslashes on Windows, URI throws an exception
-        addLocalJarFile(new File(path))
+        (addLocalJarFile(new File(path)), "local")
       } else {
         val uri = new Path(path).toUri
         // SPARK-17650: Make sure this is a valid URL before adding it to the list of dependencies
         Utils.validateURL(uri)
-        uri.getScheme match {
+        val uriSchema = uri.getScheme
+        val jarPaths = uriSchema match {
           // A JAR file which exists only on the driver node
           case null =>
             // SPARK-22585 path without schema is not url encoded
             addLocalJarFile(new File(uri.getPath))
           // A JAR file which exists only on the driver node
           case "file" => addLocalJarFile(new File(uri.getPath))
           // A JAR file which exists locally on every worker node
-          case "local" => "file:" + uri.getPath
+          case "local" => Seq("file:" + uri.getPath)
+          case "ivy" =>
+            // Since `new Path(path).toUri` will lose query information,
+            // so here we use `URI.create(path)`
+            DependencyUtils.resolveMavenDependencies(URI.create(path))
+              .map(jar => env.rpcEnv.fileServer.addJar(new File(jar)))
```

Comment:
Should we use the existing method for the sake of consistency:
```
.map(jar => addLocalJarFile(new File(jar))
```
Core logic will be the same, but there are some additional checks (verifies that `DependencyUtils` did its job correctly) and consolidates the logic.

Reply:
> Should we use the existing method for the sake of consistency:
> 
> ```
> .map(jar => addLocalJarFile(new File(jar))
> ```
> 
> Core logic will be the same, but there are some additional checks (verifies that `DependencyUtils` did its job correctly) and consolidates the logic.

should be `flatmap`, sorry for late reply, code change and conflict resolved.

Reply:
Looks good to me, thanks!